File: ReverseLinkedList2.cpp
// see : https://leetcode.com/problems/reverse-linked-list-ii/
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

// 일단 못풀었는데 답봐도 이해안됨 ㅠ
class Solution {
public:
    ListNode* reverseBetween(ListNode* head, int m, int n) {
       ListNode *dummy = new ListNode(0), *pre = dummy, *cur;
       dummy -> next = head;
        // m번째의 이전 위치
       for (int i = 0; i < m - 1; i++) {
           pre = pre -> next;
       }
        
        // m번째 위치
       cur = pre -> next;
        
        // 사이 간격만큼 반복
       for (int i = 0; i < n - m; i++) {
           ListNode* temp = pre -> next;
           pre -> next = cur -> next;
           cur -> next = cur -> next -> next;
           pre -> next -> next = temp;
       }
       return dummy -> next;
   }
};

File: Reverse_Linked_List_2.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode reverseBetween(ListNode head, int m, int n) {
        ListNode prev_m_node = new ListNode();
        ListNode m_node = new ListNode();
        ListNode cur_node = head;
        ListNode prev_node = new ListNode();
        ListNode temp = new ListNode();
        for(int num=1; num<=n; num++){
            if(num<=m){
                if(num==m-1){
                    prev_m_node = cur_node;
                }
                if(num==m){
                    m_node = cur_node;
                }
                prev_node = cur_node;
                cur_node = cur_node.next;
            }
            else{
                temp = cur_node.next;
                cur_node.next = prev_node;
                prev_node = cur_node;
                cur_node = temp;
                if(num==n){
                    prev_m_node.next = prev_node;
                    m_node.next = cur_node;
                }
                if(m == 1)
                    head = prev_node;
            }
        }
        return head;
    }
}
